crates/uv-resolver/src/error.rs

```diff
@@ -286,6 +293,62 @@ fn collapse_unavailable_workspace_members(
     }
 }
 
+/// Given a [`DerivationTree`], drop dependency incompatibilities from the root
+/// to the project.
+///
+/// Intended to effectively change the root to a workspace member in single project
+/// workspaces, avoiding a level of indirection like "And because your project
+/// requires your project, we can conclude that your projects's requirements are
+/// unsatisfiable."
+fn drop_root_dependency_on_project(
+    tree: &mut DerivationTree<PubGrubPackage, Range<Version>, UnavailableReason>,
+    project: &PackageName,
+) {
+    match tree {
+        DerivationTree::External(_) => {}
+        DerivationTree::Derived(derived) => {
```

Comment:
Yeah, I'd previously had a hard time grokking PubGrub's `collapse_no_versions` so I thought this would be harder but it wasn't too bad and I think it could be a very powerful approach in the future.